include/revng/ABI/FunctionType/Support.h

```diff
@@ -63,4 +63,18 @@ filterTypes(TrackingSortedVector<UpcastablePointer<model::Type>> &Types) {
   return Result;
 }
 
+/// Make a standalone type by placing it into a brand new `Binary`.
+///
+/// \returns A pair consisting of an unqualified `model::QualifiedType`
+///          and the `Binary` that controls the lifetime of it.
+template<derived_from<model::Type> NewType, typename... ArgumentTypes>
+[[nodiscard]] std::pair<model::QualifiedType, TupleTree<model::Binary>>
+temporaryQualifiedType(ArgumentTypes &&...Args) {
+  std::pair<model::QualifiedType, TupleTree<model::Binary>> R;
+  auto P = R.second->makeType<NewType>(std::forward<ArgumentTypes>(Args)...);
+  auto [New, NewPath] = P;
+  R.first = { NewPath, {} };
+  return R;
+}
```

Comment:
I'll gladly use anything else if you have any ideas for a way to make a useable `model::QualifiedType` without attaching it to a binary. Without something like this, I'll have to provide a bunch of overload for `model::Type` all over the place, which would suck big time.

Reply:
Please explain the use case, give me the general pictures about how `temporaryQualifiedType` fits in the `canAddAsIs` business, which is still unclear to me.

Reply:
`canAddAsIs` thing is just a wrapper around multiple checks. The check in question is the `predict` one. Which is, basically, a way to feed the specifics of `ToRaw` conversion into here to make sure we don't emit anything incompatible, no matter the inputs.

In short, if we cannot represent an arbitrary type (in this case a newly created one) as a `model::QualifiedType`, to keep similar behaviour we'd need to redesign most of `ToRaw` side to _also_ work with `model::Types`.

Reply:
By "most" here I mean `distributeAnArgument` and everything it uses/relies on.

Reply:
Ping.

Reply:
Ping?

Reply:
`temporaryQualifiedType` seems to have an overlap with `TypeBucket`. The following happens:

1. We create a type in a temporary model
2. We decide whether we want to keep it or not
3. If we want to keep it we recreate the same type in the actual model
4. We discard the temporary model

This can be done with a dedicated `TypeBucket`.

Creating a temporary model that, AFAIU, doesn't verify is not great. Reusing `TypeBucket` seems like a cleaner solution.

Reply:
That also works, I guess. It does sound like more complexity (since `TypeBucket` is going to get discarded right away), but not that much.

Thanks for the idea!

Reply:
Done: please take a look. I kept the interface thought while moving it to `TypeBucket.h`. Please take a look.

Reply:
1. I'd rename it to `makeTemporaryType` so it's clear that works as the other 2 `makeType`s we have.
2. At this point I'd just return a `std::optional`-like object that you can `*` to obtain `QualifiedType`.
3. We're still creating an empty model, which will not verify. The key point of my suggestion was to use the existing model, not a new (temporary) one.

Reply:
1. 👌
2. What do you mean? The reason why it has to be a pair is because the lifetime of the whole construct needs to be chained somewhere (second element of the pair).
3. Sadly that's the artifact of how we obtain type references. Consider that that binary object is never used for anything but `getTypePath` and `getPrimitiveType`. So it's either something like this or making these methods static. Or I guess relying on an implementation detail and calling `TypePath`'s methods directly - but I'd rather not do that.

Reply:
Please review the last commit.

Reply:
Can we avoid the `std::variant`? We can keep the class mostly as it is and just have a `merge` method such as the following:

```c++
void merge(TypeBucket &&Other) {
  revng_assert(NextAvailableTypeID == Other.FirstAvailableTypeID);
  // TODO: steal types from Other and put them in this
}
```

This ensures in the original `TypeBucket` we didn't add anything and we don't have to sprinkle `std::visitor` everywhere.

Also, `makeRelatedBucket` and `makeUnrelatedBucket` are confusing.

We have a property we want to enforce, i.e., the *original* `TypeBucket` did not receive any new type.
The proposed solution tries to enforce it in a "global" fashion while it could be verified in a `TypeBucket::merge` method, reducing how invasive it is to ensure this property holds.

Reply:
> `merge`

That's what `commit` does. Except it's even safer because this way, the "merge" is only possible into a bucket this one was explicitly spawned from.

I feel real FUD about letting any bucket just accept any other one for merge, even if the IDs make sense and the binary is the same exact same object (we can compare pointers as I'm doing here).

> We have a property we want to enforce, i.e., the original `TypeBucket` did not receive any new type.

It's pretty much doing exactly the same. Except it's in the `commit` method instead of `merge` one.

If it's `variant` you're against, I can live with `std::optional MaybeBinary, MaybeBucket;` pair, but IMO that would suck even more - since we'd have to check them all the time.

> `makeRelatedBucket` and `makeUnrelatedBucket` are kinda confusing

Yeah, I couldn't come up with good names.

- `makeRelatedBucket` is a "child"-like entity - its `commit` will merge stuff into this bucket.
- `makeUnrelatedBucket` is a "sibling"-like entity - its `commit` will merge stuff into the parent (useful in the cases where you need to either merge A or B.

Reply:
> I feel real FUD about letting any bucket just accept any other one for merge, even if the IDs make sense and the binary is the same exact same object (we can compare pointers as I'm doing here).

This is too much extra complexity for FUD. Either you can describe a credible scenario we're trying to prevent or, looking at the patchset, the extra complexity is not justified.

I think we're overfocusing on scenarios that will never happen.
There are other situations we're leaving out, I think, such as adding a type in the original model after constructing a `TypeBucket` but before `TypeBucket::commit`.

`revng_assert(NextAvailableTypeID == Other.FirstAvailableTypeID)` is good enough and keeps `TypeBucket` simple. The current solution turns `TypeBucket` into a class with two faces that could well be two distinct `class`es with a similar API.

Reply:
Pushed.

Reply:
Looks good! Can you squash?
Then I think we're good to go.

Reply:
Reopenning this thread so you don't try to merge this PR.

The solution we ended up with is broken!

It looks like there's not much choice but to go around `QualifiedType`s in this case.

Reply:
OK, can you elaborate?

Reply:
> to go around QualifiedTypes in this case

What do you mean?

I'm confused about the problem and the solution you have in mind.

Maybe you can pass around the `StructType` so we don't need to "dereference" a `TypePath`?

Reply:
Yeah, it looks like we'll have to provide at least one other overload (we still need the `QualifiedType` one), which is what I was trying to avoid.

Either way, it won't be hard to fix. I did a couple of extra tests and noticed this before it went in and screwed up something for somebody else.

Reply:
Nah, let's drop merge. Let's not have unused features.